pkg/controllers/failure_policy.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+	"sigs.k8s.io/jobset/pkg/constants"
+)
+
+// actionFunctionMap relates jobset failure policy action names to the appropriate behavior during jobset reconciliation.
+var actionFunctionMap = map[jobset.FailurePolicyAction]failurePolicyActionApplier{
+	jobset.FailJobSet:                        failJobSetActionApplier,
+	jobset.RestartJobSet:                     restartJobSetActionApplier,
+	jobset.RestartJobSetAndIgnoreMaxRestarts: restartJobSetAndIgnoreMaxRestartsActionApplier,
+}
+
+// executeFailurePolicy applies the Failure Policy of a JobSet when a failed child Job is found.
+// This function is run only when a failed child job has already been found.
+func executeFailurePolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs, updateStatusOpts *statusUpdateOpts) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If no failure policy is defined, mark the JobSet as failed.
+	if js.Spec.FailurePolicy == nil {
+		// firstFailedJob is only computed if necessary since it is expensive to compute
+		// for JobSets with many child jobs. This is why we don't unconditionally compute
+		// it once at the beginning of the function and share the results between the different
+		// possible code paths here.
+		firstFailedJob := findFirstFailedJob(ownedJobs.failed)
+		msg := messageWithFirstFailedJob(constants.FailedJobsMessage, firstFailedJob.Name)
+		setJobSetFailedCondition(ctx, js, constants.FailedJobsReason, msg, updateStatusOpts)
+		return nil
+	}
+
+	// Check for matching Failure Policy Rule
+	rules := js.Spec.FailurePolicy.Rules
+	failurePolicyRule, matchingFailedJob := findFirstFailedPolicyRuleAndJob(ctx, rules, ownedJobs.failed)
+
+	var failurePolicyRuleAction jobset.FailurePolicyAction
+	if failurePolicyRule == nil {
+		failurePolicyRuleAction = jobset.RestartJobSet
+		matchingFailedJob = findFirstFailedJob(ownedJobs.failed)
+	} else {
+		failurePolicyRuleAction = failurePolicyRule.Action
+	}
```

Comment:
```suggestion
        failurePolicyRuleAction := failurePolicyRule.Action
	if failurePolicyRule == nil {
		failurePolicyRuleAction = jobset.RestartJobSet
		matchingFailedJob = findFirstFailedJob(ownedJobs.failed)
	}
```

Reply:
The value of the pointer `failurePolicyRule` may be `nil` if there is not a matching failure policy rule. My understanding is that if `failurePolicyRule` is `nil`, then the line

``` 
failurePolicyRuleAction := failurePolicyRule.Action
```
will cause an error. Is this not correct?

Reply:
ops, missed that, sg.